storm-core/src/clj/org/apache/storm/daemon/supervisor.clj

```diff
@@ -415,23 +427,35 @@
     ;; 6. wait for workers launch
 
     (log-debug "Syncing processes")
+    (log-debug "Keepers: " keepers)
+    (log-debug "Keep ports: " keep-ports)
+    (log-debug "Reassigned executors: " reassign-executors)
     (log-debug "Assigned executors: " assigned-executors)
     (log-debug "Allocated: " allocated)
     (doseq [[id [state heartbeat]] allocated]
-      (when (not= :valid state)
-        (log-message
-         "Shutting down and clearing state for id " id
-         ". Current supervisor time: " now
-         ". State: " state
-         ", Heartbeat: " (pr-str heartbeat))
-        (shutdown-worker supervisor id)))
+      (let
+        [worker-launchtime (:launchtime (@(:worker-launchtime-atom supervisor) id))]
+        (when
+          (or 
+            (and (not= :valid state)
+                 (not= :not-started state))
+            (and (= :not-started state)
+                 (or (nil? worker-launchtime)
+                   (is-worker-launchtime-timed-out? now worker-launchtime conf))))
+              (if (= :not-started state)
+                (log-message "Worker " id " failed to start"))
+              (log-message
+                "Shutting down and clearing state for id " id
+                ". Current supervisor time: " now
+                ". State: " state
+                ", Heartbeat: " (pr-str heartbeat))
+              (shutdown-worker supervisor id))))
     (let [valid-new-worker-ids (get-valid-new-worker-ids conf supervisor reassign-executors new-worker-ids)]
       (ls-approved-workers! local-state
                         (merge
                           (select-keys (ls-approved-workers local-state)
                             (keys keepers))
-                          valid-new-worker-ids))
-      (wait-for-workers-launch conf (keys valid-new-worker-ids)))))
```

Comment:
since we removed this, we should also remove the wait-for-workers-launch private functions.